src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,327 @@
-                                    },
-                                ));
-                        }
-                    }
+                    Ok(Ok(Ok(()))) => break,
+                    Ok(Ok(Err(actual_upper))) => return Err(actual_upper),
+                    Ok(Err(invalid_use)) => panic!("{invalid_use:?}"),
+                    Err(_external_err) => continue,
                 }
             }
+            self.sync(&new_upper).await;
+        }
+        Ok(())
+    }
+
+    /// Calling this method promises that all future timestamping requests will be for offsets
+    /// greater or equal to the provided source compaction frontier. This gives permission to the
+    /// reclock operator to compact and consolidate its internal state accordingly.
+    ///
+    /// If the given source since frontier results in a change of the dest since frontier it will
+    /// be returned to the caller who can then use it to downgrade its capability.
+    pub async fn compact(
+        &mut self,
+        source_since: &HashMap<PartitionId, MzOffset>,
+    ) -> Option<AntichainRef<'_, Timestamp>> {
+        // Translate the SourceTime compaction frontier into a DestTime compaction frontier
+        let new_since = source_since
+            .iter()
+            .map(|(pid, part_since)| {
+                self.partition_uppers(pid)
+                    .find(|(_ts, offset)| part_since < offset)
+                    .expect("attempt to compact past upper")
+                    .0
+            })
+            .min()
+            .map(Antichain::from_elem)
+            .unwrap_or_else(|| self.since.clone());
+
+        if PartialOrder::less_than(&self.since, &new_since) {
+            for bindings in self.remap_trace.values_mut() {
+                // Compact the remap trace according to the computed frontier
+                for (timestamp, _) in bindings.iter_mut() {
+                    timestamp.advance_by(new_since.borrow());
+                }
+                // And then consolidate
+                consolidation::consolidate(bindings);
+            }
+            self.since = new_since;
+            self.read_handle.downgrade_since(self.since.clone()).await;
+            Some(self.since.borrow())
+        } else {
+            None
+        }
+    }
+
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
+        loop {
+            let mut new_ts = (self.now)();
+            new_ts -= new_ts % self.update_interval_ms;
+            if self.upper.less_than(&new_ts) {
+                break new_ts;
+            }
+            let sleep_ms = self.update_interval_ms - (new_ts % self.update_interval_ms);
+            tokio::time::sleep(Duration::from_millis(sleep_ms)).await;
```

Comment:
is this different than `tokio::time::sleep(Duration::From_millis(self.update_interval_ms))`?

Reply:
It is, yeah. If you're unlucky and you sample your `now_fn` at millisecond 999 then the code above will only sleep for the 1ms that's missing until you reach the next timestamping point whereas the one in our comment will always wait the maximum amount

Reply:
I think you can use a `tokio::time::interval` stored on self with https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Skip and just `self.interval.tick().await` here

Reply:
I tried doing that but because we're getting the time from an arbitrary `now_fn` I couldn't do what I described above. The tokio Interval has an internal clock (well, `Instant`s) that I assume has similar logic but doesn't necessarily correlate with how now_fn moves forward

Reply:
Right but `new_ts` here is already rounded down to the next-lowest `self.udpate_interval_ms` by `new_ts -= new_ts % self.update_interval_ms` yeah? so `(new_ts % self.update_interval_ms)` should always be `0`?